pkg/azuredisk/nodeserver.go

```diff
@@ -506,3 +521,117 @@ func (d *Driver) getBlockSizeBytes(devicePath string) (int64, error) {
 	}
 	return gotSizeBytes, nil
 }
+
+func (d *Driver) nodePublishEphemeralVolume(ctx context.Context, req *csi.NodePublishVolumeRequest) (*csi.NodePublishVolumeResponse, error) {
+	volumeId := req.GetVolumeId()
+
+	capacity := req.GetVolumeContext()["capacity"]
+	size, err := strconv.Atoi(strings.TrimSuffix(capacity, "Gi"))
+	if err != nil {
+		return nil, status.Error(codes.Internal, fmt.Sprintf("Unable to parse capacity %v", err))
+	}
+	if size == 0 {
+		size = defaultDiskSize
+	}
+
+	storageAccountType := req.GetVolumeContext()["storageAccountType"]
+	skuName, err := normalizeStorageAccountType(storageAccountType)
+	if err != nil {
+		return nil, err
+	}
+
+	ephemeralVolumeId := fmt.Sprintf("ephemeral-%s", volumeId)
+	volumeOptions := &azure.ManagedDiskOptions{
+		DiskName:           ephemeralVolumeId,
+		StorageAccountType: skuName,
+		ResourceGroup:      d.cloud.ResourceGroup,
+		SizeGB:             size,
+	}
+
+	klog.V(2).Infof("nodePublishEphemeralVolume: creating the managed disk(%s)", ephemeralVolumeId)
+	// create volume
+	diskURI, err := d.cloud.CreateManagedDisk(volumeOptions)
+	if err != nil {
+		if strings.Contains(err.Error(), "NotFound") {
+			return nil, status.Error(codes.NotFound, err.Error())
+		}
+		return nil, err
+	}
+	defer func() {
+		if err != nil {
```

Comment:
if there is error, it should stop and report error, not delete disk

Reply:
and I see there is already delete disk operation in `nodeUnpublishEphemeralVolume` func

Reply:
@pohly I am implementing the inline feature. Here we have a question. If the disk is not be deleted when some errors interrupt the `NodePublishVolume`, will the disk leak? Or will it be handled by k8s?

Reply:
This was discussed recently on Slack: https://kubernetes.slack.com/archives/C8EJ01Z46/p1578650595018200

The gist of that that discussion is that while Kubernetes may call `NodePublishVolume` again depending the error code that it returns, a CSI driver cannot count on that and therefore should try to clean up as much as possible after a failure and before returning. For example, if the pod gets deleted after a failed `NodePublishVolume`, then `NodeUnpublishVolume` will not be called.

Reply:
It is a detailed discussion. Thanks pretty for your guide!😊@pohly